Evolution/NNNN-map-failure.md

```diff
@@ -0,0 +1,149 @@
+# Map Failure
+
+* Proposal: [SAA-NNNN](NNNN-map-failure.md)
+* Authors: [Clive Liu](https://github.com/clive819)
+* Review Manager: TBD
+* Status: **Awaiting review**
+
+*During the review process, add the following fields as needed:*
+
+* Implementation: [apple/swift-async-algorithms#324](https://github.com/apple/swift-async-algorithms/pull/324)
+* Decision Notes: 
+* Bugs: 
+
+## Introduction
+
+The `mapFailure` function empowers developers to elegantly transform errors within asynchronous sequences, enhancing code readability and maintainability.
+
+```swift
+extension AsyncSequence {
+
+    public func mapFailure<MappedFailure: Error>(_ transform: @Sendable @escaping (Self.Failure) -> MappedFailure) -> some AsyncSequence<Self.Element, MappedFailure> {
+        AsyncMapFailureSequence(base: self, transform: transform)
+    }
+
+}
+```
+
+## Detailed design
+
+The actual implementation is quite simple actually - it's just simple do-catch block and invoking the transform closure inside the catch block - so we'll focus more on implementation decisions with regard to the compiler and OS versions difference.
+
+```swift
+extension AsyncSequence {
+
+#if compiler(>=6.0)
+    /// Converts any failure into a new error.
+    ///
+    /// - Parameter transform: A closure that takes the failure as a parameter and returns a new error.
+    /// - Returns: An asynchronous sequence that maps the error thrown into the one produced by the transform closure.
+    ///
+    /// Use the ``mapFailure(_:)`` operator when you need to replace one error type with another.
+    @available(macOS 15.0, iOS 18.0, watchOS 11.0, tvOS 18.0, visionOS 2.0, *)
+    public func mapFailure<MappedFailure: Error>(_ transform: @Sendable @escaping (Self.Failure) -> MappedFailure) -> some AsyncSequence<Self.Element, MappedFailure> {
+        AsyncMapFailureSequence(base: self, transform: transform)
+    }
+#endif
+
+    /// Converts any error into a new error.
+    ///
+    /// - Parameter transform: A closure that takes the error as a parameter and returns a new error.
+    /// - Returns: An asynchronous sequence that maps the error thrown into the one produced by the transform closure.
+    ///
+    /// Use the ``mapError(_:)`` operator when you need to replace one error type with another.
+    @available(macOS, deprecated: 15.0, renamed: "mapFailure")
+    @available(iOS, deprecated: 18.0, renamed: "mapFailure")
+    @available(watchOS, deprecated: 11.0, renamed: "mapFailure")
+    @available(tvOS, deprecated: 18.0, renamed: "mapFailure")
+    @available(visionOS, deprecated: 2.0, renamed: "mapFailure")
+    public func mapError<MappedError: Error>(_ transform: @Sendable @escaping (any Error) -> MappedError) -> AsyncMapErrorSequence<Self, MappedError> {
+        .init(base: self, transform: transform)
+    }
```

Comment:
I don't think we need this. It is okay if we just ship this new type when the compiler is `6.0` and above